contractcourt handoff 2

halseth opened the pull request:
alternative to #1678.

Fixes #1571.

cfromknecht commented on contractcourt/channel_arbitrator.go:
```diff
@@ -244,11 +260,44 @@ func (c *ChannelArbitrator) Start() error {
+				trigger = remoteCloseTrigger
+			}
+
+			log.Warnf("ChannelArbitrator(%v): detected stalled "+
```
i think this would be clearer as:
```golang
var trigger = chainTrigger
if isClosing {
    if c.state == StateCommitmenBroadcasted {
        switch closeType {
        case channeldb.LocalForceClose:
            trigger = localCloseTrigger
	case channeldb.RemoteForceClose:
	    trigger = remoteCloseTrigger
	}
    }
    log.Warnf()
}
```
the fallthroughs are harder to follow imo

halseth replied:
It's not equivalent, were you thinking
```go
var trigger = chainTrigger
if isClosing {
    if c.state == StateDefault
         || c.state == StateBroadcastCommit
         || c.state == StateCommitmenBroadcasted {
        switch closeType {
        case channeldb.LocalForceClose:
            trigger = localCloseTrigger
	case channeldb.RemoteForceClose:
	    trigger = remoteCloseTrigger
	}
    }
    log.Warnf()
}
```

cfromknecht replied:
ah right, forgot that local/remote could be delivered in first two, nvm :)

cfromknecht commented on contractcourt/channel_arbitrator.go:
```diff
@@ -1395,15 +1494,17 @@ func (c *ChannelArbitrator) channelAttendant(bestHeight int32) {
+				// state. When the state machine reaches that
+				// state, we'll log out the set of resolutions,
+				// such that they are available to fetch in
+				// that state.
```
I don't think we need a callback anymore, we can just log them directly and close the channel.

halseth replied:
I agree! Since we know that we are eventually gonna end up in `StateContractClosed`, there's not really any reason to wait before logging the contract resolutions. Not sure why we were using state callbacks in the first place? 🤔 cc @roasbeef

That should simplify a lot of the logic surrounding the state machine a great deal! :tada:

cfromknecht commented on contractcourt/channel_arbitrator.go:
```diff
@@ -1443,24 +1594,23 @@ func (c *ChannelArbitrator) channelAttendant(bestHeight int32) {
+				// state. When the state machine reaches that
+				// state, we'll log out the set of resolutions,
+				// such that they are available to fetch in
+				// that state.
```
similar comment here wrt. removing callback

cfromknecht commented on contractcourt/channel_arbitrator.go:
```diff
@@ -667,18 +716,38 @@ func (c *ChannelArbitrator) advanceState(triggerHeight uint32,
 		// exit early.
 		if stateCallback != nil {
 			if err := stateCallback(nextState); err != nil {
+				log.Errorf("ChannelArbitrator(%v): unable to "+
```
we can get rid of state callbacks altogether

cfromknecht commented on contractcourt/channel_arbitrator.go:
```diff
@@ -244,11 +260,44 @@ func (c *ChannelArbitrator) Start() error {
 	startingState := c.state
 	nextState, _, err := c.advanceState(
-		uint32(bestHeight), chainTrigger, nil,
+		uint32(bestHeight), trigger, nil,
```
do we always want to pass in `bestHeight` if it's not a `chainTrigger`? may need to pass in spend height?

cfromknecht commented on contractcourt/channel_arbitrator.go:
```diff
@@ -103,6 +116,9 @@ type ChannelArbitratorConfig struct {
 	// TODO(roasbeef): need RPC's to combine for pendingchannels RPC
 	MarkChannelResolved func() error
 
+	// only used for testing.
```
godoc :)

cfromknecht commented on contractcourt/channel_arbitrator_test.go:
```diff
@@ -118,17 +129,41 @@ func TestChannelArbitratorCooperativeClose(t *testing.T) {
 	assertState(t, chanArb, StateDefault)
 }
 
+func assertStateTransitions(t *testing.T, newStates <-chan ArbitratorState,
```
nice helper method!

halseth commented on contractcourt/channel_arbitrator.go:
```diff
@@ -1365,15 +1434,45 @@ func (c *ChannelArbitrator) channelAttendant(bestHeight int32) {
+
+			// Since all outputs are imemdiately available in a
+			// cooperative close, we can mark the channel resolved
+			// immediately.
```
I think I was wrong when saying it was a noop, as it actually cleans up some state :P

cfromknecht replied:
in that case we probably do want to make sure this transitions to `StateFullyResolved`, otherwise we'd need logic to force this transition on restar

cfromknecht commented on contractcourt/channel_arbitrator.go:
```diff
@@ -654,31 +700,39 @@ func (c *ChannelArbitrator) advanceState(triggerHeight uint32,
-		// If we have a state callback, then we'll attempt to execute
-		// it. If the callback doesn't execute successfully, then we'll
-		// exit early.
-		if stateCallback != nil {
```
👍

halseth commented on contractcourt/chain_arbitrator.go:
```diff
@@ -243,8 +243,22 @@ func newActiveChannelArbitrator(channel *channeldb.OpenChannel,
-		ChainArbitratorConfig:     c.cfg,
-		ChainEvents:               chanEvents,
+		MarkChannelClosed:         channel.CloseChannel,
+		IsPendingClose: func() (bool, channeldb.ClosureType, error) {
```
I made it return `true` or `false` depending on what we already now when creating the config, noting in the godoc of `IsPendingClose` that the value is only valid at startup.

halseth commented:
Addressed comments and added more unit test coverage. PTAL.

Regarding saving the DBs of users that already have had their channels closed without resolutions: it doesn't look like `ChannelClosSummary` contain everything we need from the `OpenChannel` struct to directly recover the funds. But maybe some of it can be derived from what we store?

cfromknecht commented on contractcourt/channel_arbitrator.go:
```diff
@@ -1365,15 +1414,54 @@ func (c *ChannelArbitrator) channelAttendant(bestHeight int32) {
+			}
+
+			// To clean up, we ensure we commit to StateFullyResolved.
+			c.state = StateFullyResolved
```
What is the rationale for reimplementing channel resolved + commit state here, instead of using `advanceState`? The latter seems more fault tolerant as rn idt we will properly recover if either the `Closed -> Resolved` or `Resolved -> State Committed` transitions fail

cfromknecht replied:
we won't though, because the channel has already been closed. we only resubscribe for open channels.

halseth replied:
I agree that it would be cleanest to make coop closes also go through the state machine. It would require a new trigger type, and probably also writing (empty) resolutions for the coop close case. Wouldn't be that big of a change though, and would be nice to have all closes (except breaches) follow the same logic :)

Regarding replaying of this state, it is benign, as after the coop close has been called with `CloseChannel` all the necessary state is written in the database, `MarkChannelResolved` will only clean up some in-memory state, and `CommitState` will write the state to the DB, but it will never be read again.

Let me know what you think!

halseth replied:
After @cfromknecht's suggestion I added a new trigger `coopCloseTrigger` and make the state machine skip ahead to `StateFullyResolved` in case it triggers. Very happy with the result, a lot cleaner now! 👍

cfromknecht commented on contractcourt/channel_arbitrator_test.go:
```diff
@@ -118,21 +196,75 @@ func TestChannelArbitratorCooperativeClose(t *testing.T) {
+func assertStateTransitions(t *testing.T, newStates <-chan ArbitratorState,
+	expectedStates ...ArbitratorState) {
+
+	_, _, line, _ := runtime.Caller(1)
```
use `t.Helper()`?

cfromknecht commented on contractcourt/channel_arbitrator_test.go:
```diff
@@ -2,16 +2,91 @@ package contractcourt
 	"github.com/lightningnetwork/lnd/lnwire"
 )
 
+type mockArbitratorLog struct {
```
woo!

Roasbeef commented:
> Regarding saving the DBs of users that already have had their channels closed without resolutions: 

In this case, perhaps we can just wait for them to make an issue or bring this up on slack or w/e, and we can provide a patch to manually transition them? The closed summary in the database contains the key state of the last commit (well it only recently does for _all_ closes...) so we can manually re-create the resolutions if needed. This allows us to not have a special case in the code that will only ever be executed _once_ by a hand full of people.

Roasbeef commented on contractcourt/chain_arbitrator.go:
```diff
@@ -398,6 +403,11 @@ func (c *ChainArbitrator) Start() error {
 			ChainEvents:           &ChainEventSubscription{},
+			IsPendingClose: func() (bool, uint32,
+				channeldb.ClosureType, error) {
+				return true, closeChanInfo.CloseHeight,
```
Yep, gotta be aware of loop variable binding. We can do something like this above:
````
closeHeight := closeChanInfo.CloseHeight
````

cfromknecht replied:
seems cleaner just to set fields tbh, lot of extra code for values that are o/w immutable

halseth replied:
Now just setting fields, much cleaner :)

Roasbeef commented on contractcourt/channel_arbitrator.go:
```diff
@@ -95,6 +95,21 @@ type ChannelArbitratorConfig struct {
+	//
+	// NOTE: This value is currently only valid at startup, and won't
+	// change after the ChannelArbitrator is first created.
+	IsPendingClose func() (bool, uint32, channeldb.ClosureType, error)
```
Doesn't seem like it.

Roasbeef commented on contractcourt/channel_arbitrator.go:
```diff
@@ -244,12 +259,45 @@ func (c *ChannelArbitrator) Start() error {
+			case channeldb.RemoteForceClose:
+				trigger = remoteCloseTrigger
+			}
+			triggerHeight = closeHeight
```
👍 

This should actually resolve a known bug in the case of neutrino.

cfromknecht replied:
would we start searching from most recent block height instead?

Roasbeef replied:
As in, if best height is 10, but then the event happened at height 7, we would set the wrong height hint.

cfromknecht replied:
didn't know that was an existing issue, great that we have a fix 🎉

halseth commented on contractcourt/channel_arbitrator.go:
```diff
@@ -613,7 +613,11 @@ func (c *ChannelArbitrator) stateStep(triggerHeight uint32,
 			"on-chain at height=%v", c.cfg.ChanPoint, triggerHeight)
-		return nextState, closeTx, c.cfg.MarkChannelResolved()
+
+		if err := c.cfg.MarkChannelResolved(); err != nil {
```
It is to advance the state machine to the `StateFullyResolved` state. It won't matter much, as we'll never restart the state machine for a resolved channel, but it makes the behaviour consistent for all states.

halseth commented:
Squashed, rebased and addressed comments.

Also added a check+test for the case where the channel is marked closed in the DB, but resolutions are not found. As this would previously cause the `channelArbitrator` to fail at startup, we now ignore the error to not prevent `lnd` startup. This case must be handled on an individual basis.

One question remains, whether to also thread coop closes through the state machine. For consistency I think we should.

cfromknecht commented on contractcourt/channel_arbitrator.go:
```diff
@@ -244,15 +262,56 @@ func (c *ChannelArbitrator) Start() error {
+				trigger = localCloseTrigger
+			case channeldb.RemoteForceClose:
+				trigger = remoteCloseTrigger
+			}
```
i think we also want a case for coop closes here, so that we can handle:
1) channel closed by channel arb
2) lnd crash
3) channel still in Default, but has been closed
4) want to end up in FullyResolved

halseth replied:
Since a coop close is always marked with `CloseChannelSummary{ IsPending: false }` this won't ever be hit. 

However, do you think we should change it to `IsPending: true` for consistency, and make it be fully closed (`IsPending: false`) when `MarkChannelResolved` is called? I originally didn't make this change to be conservative and change behaviour as little as possible.

halseth replied:
Ended up setting `IsPending: true` also for coop closes.

cfromknecht commented on contractcourt/channel_arbitrator.go:
```diff
@@ -1365,16 +1438,29 @@ func (c *ChannelArbitrator) channelAttendant(bestHeight int32) {
+			_, _, err = c.advanceState(
+				closeInfo.CloseHeight, coopCloseTrigger,
+			)
+
```
nit: extra new line

halseth replied:
fin

cfromknecht commented:
we 💚
